libcudacxx/include/cuda/__floating_point/config.h

```diff
@@ -0,0 +1,526 @@
+//   - __nans(): returns the signaling NaN value for the floating point type (if __has_signaling_nan is true)
+//   - __is_inf(): returns whether the given value is infinity (if __has_inf is true)
+//   - __is_nan(): returns whether the given value is NaN (if __has_nan or __has_nans is true)
+
+struct __fp_no_native_type_tag
+{};
+
+struct __fp_invalid_config
+{};
+
+struct __fp_from_native_t
+{};
+
+_CCCL_INLINE_VAR constexpr __fp_from_native_t __fp_from_native{};
+
+struct __fp_from_storage_t
+{};
+
+_CCCL_INLINE_VAR constexpr __fp_from_storage_t __fp_from_storage{};
+
+struct __fp_iec559_config_base
+{
+  static constexpr bool __is_signed  = true;
+  static constexpr bool __has_inf    = true;
+  static constexpr bool __has_nan    = true;
+  static constexpr bool __has_nans   = true;
+  static constexpr bool __has_denorm = true;
+  static constexpr bool __is_iec559  = true;
+
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(__is_cuda_extended_floating_point_v<_Tp>)
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __min()
+  {
+    _Tp __ret{};
+    __ret.__set_exp(typename _Tp::__storage_type{1});
+    return __ret;
+  }
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(__is_cuda_extended_floating_point_v<_Tp>)
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __max()
+  {
+    _Tp __ret{};
+    __ret.__set_exp(static_cast<typename _Tp::__storage_type>(~typename _Tp::__storage_type{1}));
+    return __ret;
+  }
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(__is_cuda_extended_floating_point_v<_Tp>)
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __inf()
+  {
+    _Tp __ret{};
+    __ret.__set_exp(_Tp::__exp_val_mask());
+    return __ret;
+  }
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(__is_cuda_extended_floating_point_v<_Tp>)
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __nan()
+  {
+    _Tp __ret = __inf<_Tp>();
+    __ret.__set_mant(typename _Tp::__storage_type{1} << (_Tp::__mant_nbits - 1));
+    return __ret;
+  }
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(__is_cuda_extended_floating_point_v<_Tp>)
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __nans()
+  {
+    _Tp __ret = __inf<_Tp>();
+    __ret.__set_mant(typename _Tp::__storage_type{1} << (_Tp::__mant_nbits - 2));
+    return __ret;
+  }
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(__is_cuda_extended_floating_point_v<_Tp>)
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr bool __is_inf(const _Tp& __val)
+  {
+    return __val.__get_exp() == _Tp::__exp_val_mask() && __val.__get_mant() == typename _Tp::__storage_type{0};
+  }
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(__is_cuda_extended_floating_point_v<_Tp>)
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr bool __is_nan(const _Tp& __val)
+  {
+    return __val.__get_exp() == _Tp::__exp_val_mask() && __val.__get_mant() != typename _Tp::__storage_type{0};
+  }
+};
+
+struct __fp4_e2m1_config
```

Comment:
If possible, I'd love to fold these into a template too, because it seems to me that most of the code is shared, just with specific constants being different - is that right?